core/src/main/java/hudson/model/Computer.java

```diff
@@ -1102,6 +1102,8 @@ protected void removeExecutor(final Executor e) {
                     if (ciBase != null) { // TODO confirm safe to assume non-null and use getInstance()
                         ciBase.removeComputer(Computer.this);
                     }
+                } else if (isIdle()) {
```

Comment:
Maybe this should be moved out of the `synchronized` block? Or the notification if not the `isIdle` check. Not clear if there is a possible lock ordering problem here. https://github.com/jenkinsci/jenkins/blob/5df30de4358ca93db17eb1d80a6b3a6f1c92ffad/core/src/main/java/hudson/model/Executor.java#L621-L622

Reply:
Lock ordering is the same as for isAlive() check, and I could look up either individual executor locks, or Queue > Executor lock. Computer > Executor does seem a logical order.
On the other hand, it makes sense to execute the listener outside the lock.